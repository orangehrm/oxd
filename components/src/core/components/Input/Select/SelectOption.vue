<template>
  <div @mousedown="onClick" role="option" class="oxd-select-option">
    <slot></slot>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';

export default defineComponent({
  name: 'oxd-select-option',

  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
  },

  emits: ['select'],

  methods: {
    onClick() {
      if (this.disabled) return;
      this.$emit('select');
    },
  },
});
</script>

<style src="./select-input.scss" lang="scss" scoped></style>
