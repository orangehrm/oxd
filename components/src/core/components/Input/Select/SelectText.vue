<template>
  <div
    :class="classes"
    :style="style"
    :tabindex="tabIndex"
    v-bind="defaultAttrs"
    @focus="onFocus"
    @blur="onBlur"
  >
    <div class="oxd-select-text-input">
      <div class="selected-content">{{ value }}</div>
      <input
        type="text"
        readonly="readonly"
        v-bind="inputAttrs"
        @focus="onFocus"
        @blur="$attrs.onBlur"
      />
    </div>
    <div class="oxd-select-text--after-slot-option">
      <slot name="afterInput"></slot>
    </div>
    <div class="oxd-select-text--after">
      <div :class="dropdownIconClasses">
        <oxd-icon
          v-if="!disabled"
          :class="dropdownIconClasses"
          :size="dropdownIconSize"
          :name="dropdownIcon"
        />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import eventsMixin from './events-mixin';
import navigationMixin from './navigation-mixin';

export default defineComponent({
  name: 'oxd-select-text',
  inheritAttrs: false,

  components: {
    'oxd-icon': Icon,
  },

  mixins: [navigationMixin, eventsMixin],

  emits: ['icon:clicked'],

  props: {
    dropdownOpened: {
      type: Boolean,
      default: true,
    },
    value: {
      type: String,
    },
    style: {
      type: Object,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    downIcon: {
      type: String,
      default: 'caret-down-fill',
    },
    upIcon: {
      type: String,
      default: 'caret-up-fill',
    },
    dropdownIconSize: {
      type: String,
      default: 'medium',
    },
  },

  data() {
    return {
      focused: false,
    };
  },

  computed: {
    classes(): object {
      return {
        'oxd-select-text': true,
        'oxd-select-text--active': !this.focused,
        'oxd-select-text--focus': this.focused,
        'oxd-select-text--error': this.hasError,
        'oxd-select-text--disabled': this.disabled,
        'oxd-select-text--readonly': this.readonly,
      };
    },
    dropdownIcon(): string {
      return this.dropdownOpened ? this.upIcon : this.downIcon;
    },
    dropdownIconClasses(): object {
      return {
        'oxd-select-text--arrow': true,
        '--disabled': this.disabled,
        '--readonly': this.readonly,
      };
    },
    tabIndex(): number {
      return this.disabled ? -1 : 0;
    },
    inputAttrs() {
      const allowed = ['id'];
      return Object.keys(this.$attrs)
        .filter(key => allowed.includes(key))
        .reduce((obj, key) => {
          obj[key] = this.$attrs[key];
          return obj;
        }, {});
    },
    defaultAttrs() {
      const notAllowed = ['id'];
      return Object.keys(this.$attrs)
        .filter(key => !notAllowed.includes(key))
        .reduce((obj, key) => {
          obj[key] = this.$attrs[key];
          return obj;
        }, {});
    },
  },

  methods: {
    onFocus($e: Event) {
      if (this.disabled) {
        $e.stopImmediatePropagation();
        return;
      }
      this.focused = true;
    },
    onBlur() {
      this.focused = false;
    },
  },
});
</script>

<style src="./select-input.scss" lang="scss" scoped></style>
