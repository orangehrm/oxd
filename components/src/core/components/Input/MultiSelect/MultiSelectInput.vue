<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <div class="oxd-multiselect-wrapper">
    <oxd-select-text
      v-bind="$attrs"
      :disabled="disabled"
      :readonly="readonly"
      @blur="onBlur"
      @focus="onOpenDropdown"
      @keyup.esc="onCloseDropdown"
      @keydown.enter.prevent="onSelectEnter"
      @keydown.down.exact.prevent="onSelectDown"
      @keydown.up.exact.prevent="onSelectUp"
      @keydown="onKeypress"
    >
      <template v-slot:afterInput>
        <slot
          v-if="modelValue.length > 0"
          name="afterSelected"
          :data="modelValue"
        ></slot>
      </template>
    </oxd-select-text>

    <oxd-select-dropdown
      v-if="dropdownOpen"
      :class="dropdownClasses"
      :loading="loading"
      :empty="computedOptions.length === 0"
      :empty-text="emptyText"
    >
      <oxd-select-option
        v-for="(option, i) in computedOptions"
        :key="option.id"
        :class="optionClasses[i]"
        :disabled="option._disabled || option._selected"
        :ref="`option-${i}`"
        @select="onSelect(option)"
      >
        <slot name="option" :data="option"></slot>
        <span v-if="!$slots['option']">{{ option.label }}</span>
      </oxd-select-option>
    </oxd-select-dropdown>

    <oxd-mutliselect-chips
      :disabled="disabled"
      :readonly="readonly"
      :selected="modelValue"
      @chipRemoved="onRemoveSelected"
    ></oxd-mutliselect-chips>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import eventsMixin from '../Select/events-mixin';
import navigationMixin from '../Select/navigation-mixin';
import {TOP, BOTTOM, Option, Position, DROPDOWN_POSITIONS} from '../types';
import SelectText from '@ohrm/oxd/core/components/Input/Select/SelectText.vue';
import SelectDropdown from '@ohrm/oxd/core/components/Input/Select/SelectDropdown.vue';
import SelectOption from '@ohrm/oxd/core/components/Input/Select/SelectOption.vue';
import MultiSelectChips from '@ohrm/oxd/core/components/Input/MultiSelect/MultiSelectChips.vue';

export default defineComponent({
  name: 'oxd-multiselect-input',
  inheritAttrs: false,

  components: {
    'oxd-select-text': SelectText,
    'oxd-select-dropdown': SelectDropdown,
    'oxd-select-option': SelectOption,
    'oxd-mutliselect-chips': MultiSelectChips,
  },

  mixins: [navigationMixin, eventsMixin],

  emits: [
    'update:modelValue',
    'dropdown:opened',
    'dropdown:closed',
    'dropdown:blur',
    'dropdown:clear',
  ],

  props: {
    modelValue: {
      type: Array,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    options: {
      type: Array,
      required: true,
    },
    emptyText: {
      type: String,
      required: false,
    },
    dropdownPosition: {
      type: String,
      default: BOTTOM,
      validator: function(value: Position) {
        return DROPDOWN_POSITIONS.indexOf(value) !== -1;
      },
    },
  },

  data() {
    return {
      focused: false,
      loading: false,
      dropdownOpen: false,
      searchTerm: null,
    };
  },

  computed: {
    computedOptions(): Option[] {
      return this.options.map((option: Option) => {
        let _selected = false;
        if (Array.isArray(this.modelValue)) {
          _selected = this.modelValue.findIndex(o => o.id === option.id) > -1;
        }
        return {...option, _selected};
      });
    },
    dropdownClasses(): object {
      return {
        '--positon-bottom': this.dropdownPosition === BOTTOM,
        '--positon-top': this.dropdownPosition === TOP,
      };
    },
    optionClasses(): object[] {
      return this.computedOptions.map((option: Option, index: number) => {
        return {
          '--disabled': option._disabled,
          '--selected': option._selected,
          '--focused': index === this.pointer,
          [`--indent-${option._indent}`]: option._indent !== undefined,
        };
      });
    },
  },

  watch: {
    pointer(newIndex: number) {
      const option = this.$refs[`option-${newIndex}`];
      if (option?.$el) this.scrollToView(option.$el);
    },
  },
});
</script>

<style src="./multiselect-input.scss" lang="scss" scoped></style>
