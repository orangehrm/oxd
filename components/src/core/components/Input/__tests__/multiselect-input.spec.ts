import {mount} from '@vue/test-utils';
import MultiSelectInput from '@orangehrm/oxd/core/components/Input/MultiSelect/MultiSelectInput.vue';
import SelectText from '@orangehrm/oxd/core/components/Input/Select/SelectText.vue';
import SelectOption from '@orangehrm/oxd/core/components/Input/Select/SelectOption.vue';

const options = [
  {
    id: 1,
    label: 'HR Admin',
  },
  {
    id: 2,
    label: 'ESS User',
  },
  {
    id: 3,
    label: 'Supervisor',
  },
];

describe('MultiSelectInput.vue', () => {
  it('renders OXD MultiSelect Input', () => {
    const wrapper = mount(MultiSelectInput, {
      props: {options, modelValue: []},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should load options to Select', async () => {
    const wrapper = mount(MultiSelectInput, {
      props: {options, modelValue: []},
    });
    wrapper.findComponent(SelectText).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    expect(nodes.length).toBe(3);
  });
  it('should select one option', async () => {
    const wrapper = mount(MultiSelectInput, {
      props: {options, modelValue: []},
    });
    wrapper.findComponent(SelectText).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    await nodes[0].trigger('mousedown');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [
        [
          {
            id: 1,
            label: 'HR Admin',
            _selected: false,
          },
        ],
      ],
    ]);
  });
  it('should not select already selected option', async () => {
    const wrapper = mount(MultiSelectInput, {
      props: {
        options,
        modelValue: [
          {
            id: 1,
            label: 'HR Admin',
          },
        ],
      },
    });
    wrapper.findComponent(SelectText).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    await nodes[0].trigger('mousedown');
    expect(wrapper.emitted('update:modelValue')).toBeFalsy();
  });
  it('should be able to select multiple options', async () => {
    const wrapper = mount(MultiSelectInput, {
      props: {
        options,
        modelValue: [
          {
            id: 1,
            label: 'HR Admin',
          },
        ],
      },
    });
    wrapper.findComponent(SelectText).trigger('click');
    await wrapper.vm.$nextTick();
    const nodes = wrapper.findAllComponents(SelectOption);
    await nodes[1].trigger('mousedown');
    expect(wrapper.emitted('update:modelValue')).toEqual([
      [
        [
          {
            id: 1,
            label: 'HR Admin',
          },
          {
            id: 2,
            label: 'ESS User',
            _selected: false,
          },
        ],
      ],
    ]);
  });
  it('should able to remove selected chip', async () => {
    const wrapper = mount(MultiSelectInput, {
      props: {
        options,
        modelValue: [
          {
            id: 1,
            label: 'HR Admin',
          },
        ],
      },
    });
    const cancelButton = wrapper.find("[data-test='removeIcon']");
    await cancelButton.trigger('click');
    expect(wrapper.emitted()).toHaveProperty('update:modelValue');
    expect(wrapper.emitted('update:modelValue')).toEqual([[[]]]);
  });
  it('should able to remove selected chip when readonly mode', async () => {
    const wrapper = mount(MultiSelectInput, {
      props: {
        options,
        readonly: true,
        modelValue: [
          {
            id: 1,
            label: 'HR Admin',
          },
        ],
      },
    });
    const cancelButton = wrapper.find("[data-test='removeIcon']");
    await cancelButton.trigger('click');
    expect(wrapper.emitted('update:modelValue')).toBeFalsy();
  });
  it('should able to remove selected chip when disabled mode', async () => {
    const wrapper = mount(MultiSelectInput, {
      props: {
        options,
        disabled: true,
        modelValue: [
          {
            id: 1,
            label: 'HR Admin',
          },
        ],
      },
    });
    const cancelButton = wrapper.find("[data-test='removeIcon']");
    await cancelButton.trigger('click');
    expect(wrapper.emitted('update:modelValue')).toBeFalsy();
  });
});
