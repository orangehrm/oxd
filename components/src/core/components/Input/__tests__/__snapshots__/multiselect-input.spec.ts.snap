// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectInput.vue renders OXD MultiSelect Input 1`] = `
<div class="oxd-multiselect-wrapper">
  <div class="oxd-select-text oxd-select-text--active" tabindex="0">
    <div class="oxd-select-text-input">
      <div class="selected-content"></div><input type="text" readonly="">
    </div>
    <div class="oxd-select-text--after">
      <div class="oxd-select-text--arrow"><i class="oxd-icon oxd-icon--medium bi-caret-down-fill"></i></div>
    </div>
  </div>
  <!--v-if-->
  <div class="oxd-multiselect-chips-area"></div>
</div>
`;
